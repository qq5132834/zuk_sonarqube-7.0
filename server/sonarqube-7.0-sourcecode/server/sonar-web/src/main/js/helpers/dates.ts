/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import {
  differenceInDays as _differenceInDays,
  differenceInSeconds as _differenceInSeconds,
  differenceInYears as _differenceInYears,
  isSameDay as _isSameDay,
  parse,
  startOfDay as _startOfDay
} from 'date-fns';

function pad(number: number) {
  if (number < 10) {
    return '0' + number;
  }
  return number;
}

type ParsableDate = string | number | Date;

export function parseDate(rawDate: ParsableDate): Date {
  return parse(rawDate);
}

export function toShortNotSoISOString(rawDate: ParsableDate): string {
  const date = parseDate(rawDate);
  return `${date.getFullYear()}-${pad(date.getMonth() + 1)}-${pad(date.getDate())}`;
}

export function toNotSoISOString(rawDate: ParsableDate): string {
  const date = parseDate(rawDate);
  return date.toISOString().replace(/\..+Z$/, '+0000');
}

export function startOfDay(date: Date): Date {
  return _startOfDay(date);
}

export function isValidDate(date: Date): boolean {
  return !isNaN(date.getTime());
}

export function isSameDay(dateLeft: Date, dateRight: Date): boolean {
  return _isSameDay(dateLeft, dateRight);
}

export function differenceInYears(dateLeft: Date, dateRight: Date): number {
  return _differenceInYears(dateLeft, dateRight);
}

export function differenceInDays(dateLeft: Date, dateRight: Date): number {
  return _differenceInDays(dateLeft, dateRight);
}

export function differenceInSeconds(dateLeft: Date, dateRight: Date): number {
  return _differenceInSeconds(dateLeft, dateRight);
}
