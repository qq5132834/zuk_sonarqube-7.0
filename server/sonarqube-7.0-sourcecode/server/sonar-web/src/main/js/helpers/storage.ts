/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
const PROJECTS_DEFAULT_FILTER = 'sonarqube.projects.default';
const PROJECTS_FAVORITE = 'favorite';
const PROJECTS_ALL = 'all';

const PROJECTS_VIEW = 'sonarqube.projects.view';
const PROJECTS_VISUALIZATION = 'sonarqube.projects.visualization';
const PROJECTS_SORT = 'sonarqube.projects.sort';

const PROJECT_ACTIVITY_GRAPH = 'sonarqube.project_activity.graph';
const PROJECT_ACTIVITY_GRAPH_CUSTOM = 'sonarqube.project_activity.graph.custom';

function save(key: string, value?: string): void {
  try {
    if (value) {
      window.localStorage.setItem(key, value);
    } else {
      window.localStorage.removeItem(key);
    }
  } catch (e) {
    // usually that means the storage is full
    // just do nothing in this case
  }
}

export function saveFavorite(): void {
  save(PROJECTS_DEFAULT_FILTER, PROJECTS_FAVORITE);
}

export function isFavoriteSet(): boolean {
  const setting = window.localStorage.getItem(PROJECTS_DEFAULT_FILTER);
  return setting === PROJECTS_FAVORITE;
}

export function saveAll(): void {
  save(PROJECTS_DEFAULT_FILTER, PROJECTS_ALL);
}

export function isAllSet(): boolean {
  const setting = window.localStorage.getItem(PROJECTS_DEFAULT_FILTER);
  return setting === PROJECTS_ALL;
}

export function saveView(view?: string): void {
  save(PROJECTS_VIEW, view);
}

export function getView(): string | null {
  return window.localStorage.getItem(PROJECTS_VIEW);
}

export function saveVisualization(visualization?: string): void {
  save(PROJECTS_VISUALIZATION, visualization);
}

export function getVisualization(): string | null {
  return window.localStorage.getItem(PROJECTS_VISUALIZATION);
}

export function saveSort(sort?: string): void {
  save(PROJECTS_SORT, sort);
}

export function getSort(): string | null {
  return window.localStorage.getItem(PROJECTS_SORT);
}

export function saveCustomGraph(metrics?: string[]): void {
  save(PROJECT_ACTIVITY_GRAPH_CUSTOM, metrics ? metrics.join(',') : '');
}

export function getCustomGraph(): string[] {
  const customGraphs = window.localStorage.getItem(PROJECT_ACTIVITY_GRAPH_CUSTOM);
  return customGraphs ? customGraphs.split(',') : [];
}

export function saveGraph(graph?: string): void {
  save(PROJECT_ACTIVITY_GRAPH, graph);
}

export function getGraph(): string {
  return window.localStorage.getItem(PROJECT_ACTIVITY_GRAPH) || 'issues';
}
