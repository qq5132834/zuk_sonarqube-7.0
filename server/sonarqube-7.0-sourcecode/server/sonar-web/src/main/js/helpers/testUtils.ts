/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { shallow, ShallowRendererProps, ShallowWrapper, ReactWrapper } from 'enzyme';
import { IntlProvider } from 'react-intl';

export const mockEvent = {
  target: { blur() {} },
  currentTarget: { blur() {} },
  preventDefault() {},
  stopPropagation() {}
};

export function click(element: ShallowWrapper | ReactWrapper, event = {}): void {
  element.simulate('click', { ...mockEvent, ...event });
}

export function clickOutside(event = {}): void {
  const dispatchedEvent = new MouseEvent('click', event);
  window.dispatchEvent(dispatchedEvent);
}

export function submit(element: ShallowWrapper): void {
  element.simulate('submit', {
    preventDefault() {}
  });
}

export function change(element: ShallowWrapper, value: string, event = {}): void {
  element.simulate('change', {
    target: { value },
    currentTarget: { value },
    ...event
  });
}

export function keydown(keyCode: number): void {
  const event = new KeyboardEvent('keydown', { keyCode } as KeyboardEventInit);
  document.dispatchEvent(event);
}

export function elementKeydown(element: ShallowWrapper, keyCode: number): void {
  const event = {
    currentTarget: { element },
    keyCode,
    preventDefault() {}
  };

  if (typeof element.type() === 'string') {
    // `type()` is string for native dom elements
    element.simulate('keydown', event);
  } else {
    element.prop<Function>('onKeyDown')(event);
  }
}

export function doAsync(fn?: Function): Promise<void> {
  return new Promise(resolve => {
    setImmediate(() => {
      if (fn) {
        fn();
      }
      resolve();
    });
  });
}

// Create the IntlProvider to retrieve context for wrapping around.
const intlProvider = new IntlProvider({ locale: 'en' }, {});
const { intl } = intlProvider.getChildContext();
export function shallowWithIntl(node: React.ReactElement<any>, options: ShallowRendererProps = {}) {
  return shallow(node, { ...options, context: { intl, ...options.context } });
}
