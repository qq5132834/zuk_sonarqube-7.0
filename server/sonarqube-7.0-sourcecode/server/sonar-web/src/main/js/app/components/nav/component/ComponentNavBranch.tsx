/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import * as classNames from 'classnames';
import * as PropTypes from 'prop-types';
import ComponentNavBranchesMenu from './ComponentNavBranchesMenu';
import SingleBranchHelperPopup from './SingleBranchHelperPopup';
import NoBranchSupportPopup from './NoBranchSupportPopup';
import { Branch, Component } from '../../../types';
import * as theme from '../../../theme';
import BranchIcon from '../../../../components/icons-components/BranchIcon';
import { isShortLivingBranch } from '../../../../helpers/branches';
import { translate } from '../../../../helpers/l10n';
import HelpIcon from '../../../../components/icons-components/HelpIcon';
import BubblePopupHelper from '../../../../components/common/BubblePopupHelper';
import Tooltip from '../../../../components/controls/Tooltip';

interface Props {
  branches: Branch[];
  component: Component;
  currentBranch: Branch;
  location?: any;
}

interface State {
  dropdownOpen: boolean;
  noBranchSupportPopupOpen: boolean;
  singleBranchPopupOpen: boolean;
}

export default class ComponentNavBranch extends React.PureComponent<Props, State> {
  mounted: boolean;

  static contextTypes = {
    branchesEnabled: PropTypes.bool.isRequired,
    onSonarCloud: PropTypes.bool
  };

  constructor(props: Props) {
    super(props);
    this.state = {
      dropdownOpen: false,
      noBranchSupportPopupOpen: false,
      singleBranchPopupOpen: false
    };
  }

  componentDidMount() {
    this.mounted = true;
  }

  componentWillReceiveProps(nextProps: Props) {
    if (
      nextProps.component !== this.props.component ||
      this.differentBranches(nextProps.currentBranch, this.props.currentBranch) ||
      nextProps.location !== this.props.location
    ) {
      this.setState({ dropdownOpen: false, singleBranchPopupOpen: false });
    }
  }

  componentWillUnmount() {
    this.mounted = false;
  }

  differentBranches(a: Branch, b: Branch) {
    // if main branch changes name, we should not close the dropdown
    return a.isMain && b.isMain ? false : a.name !== b.name;
  }

  handleClick = (event: React.SyntheticEvent<HTMLElement>) => {
    event.preventDefault();
    event.stopPropagation();
    event.currentTarget.blur();
    this.setState(state => ({ dropdownOpen: !state.dropdownOpen }));
  };

  closeDropdown = () => {
    if (this.mounted) {
      this.setState({ dropdownOpen: false });
    }
  };

  toggleSingleBranchPopup = (show?: boolean) => {
    if (show !== undefined) {
      this.setState({ singleBranchPopupOpen: show });
    } else {
      this.setState(state => ({ singleBranchPopupOpen: !state.singleBranchPopupOpen }));
    }
  };

  toggleNoBranchSupportPopup = (show?: boolean) => {
    if (show !== undefined) {
      this.setState({ noBranchSupportPopupOpen: show });
    } else {
      this.setState(state => ({ noBranchSupportPopupOpen: !state.noBranchSupportPopupOpen }));
    }
  };

  handleSingleBranchClick = (event: React.SyntheticEvent<HTMLAnchorElement>) => {
    event.preventDefault();
    event.stopPropagation();
    this.toggleSingleBranchPopup();
  };

  handleNoBranchSupportClick = (event: React.SyntheticEvent<HTMLAnchorElement>) => {
    event.preventDefault();
    event.stopPropagation();
    this.toggleNoBranchSupportPopup();
  };

  renderDropdown = () => {
    const { configuration } = this.props.component;
    return this.state.dropdownOpen ? (
      <ComponentNavBranchesMenu
        branches={this.props.branches}
        canAdmin={configuration && configuration.showSettings}
        component={this.props.component}
        currentBranch={this.props.currentBranch}
        onClose={this.closeDropdown}
      />
    ) : null;
  };

  renderMergeBranch = () => {
    const { currentBranch } = this.props;
    if (!isShortLivingBranch(currentBranch)) {
      return null;
    }
    return currentBranch.isOrphan ? (
      <span className="note big-spacer-left text-lowercase">
        {translate('branches.orphan_branch')}
        <Tooltip overlay={translate('branches.orphan_branches.tooltip')}>
          <i className="icon-help spacer-left" />
        </Tooltip>
      </span>
    ) : (
      <span className="note big-spacer-left text-lowercase">
        {translate('from')} <strong>{currentBranch.mergeBranch}</strong>
      </span>
    );
  };

  renderSingleBranchPopup = () => (
    <div className="display-inline-block spacer-left">
      <a className="link-no-underline" href="#" onClick={this.handleSingleBranchClick}>
        <HelpIcon fill={theme.blue} />
      </a>
      <BubblePopupHelper
        isOpen={this.state.singleBranchPopupOpen}
        position="bottomleft"
        popup={<SingleBranchHelperPopup />}
        togglePopup={this.toggleSingleBranchPopup}
      />
    </div>
  );

  renderNoBranchSupportPopup = () => (
    <div className="display-inline-block spacer-left">
      <a className="link-no-underline" href="#" onClick={this.handleNoBranchSupportClick}>
        <HelpIcon fill={theme.gray80} />
      </a>
      <BubblePopupHelper
        isOpen={this.state.noBranchSupportPopupOpen}
        position="bottomleft"
        popup={<NoBranchSupportPopup />}
        togglePopup={this.toggleNoBranchSupportPopup}
      />
    </div>
  );

  render() {
    const { branches, currentBranch } = this.props;

    if (this.context.onSonarCloud && !this.context.branchesEnabled) {
      return null;
    }

    if (!this.context.branchesEnabled) {
      return (
        <div className="navbar-context-branches">
          <BranchIcon branch={currentBranch} className="little-spacer-right" fill={theme.gray80} />
          <span className="note">{currentBranch.name}</span>
          {this.renderNoBranchSupportPopup()}
        </div>
      );
    }

    if (branches.length < 2) {
      return (
        <div className="navbar-context-branches">
          <BranchIcon branch={currentBranch} className="little-spacer-right" />
          <span className="note">{currentBranch.name}</span>
          {this.renderSingleBranchPopup()}
        </div>
      );
    }

    return (
      <div
        className={classNames('navbar-context-branches', 'dropdown', {
          open: this.state.dropdownOpen
        })}>
        <a className="link-base-color link-no-underline" href="#" onClick={this.handleClick}>
          <BranchIcon branch={currentBranch} className="little-spacer-right" />
          <Tooltip overlay={currentBranch.name} mouseEnterDelay={1}>
            <span className="text-limited text-top">{currentBranch.name}</span>
          </Tooltip>
          <i className="icon-dropdown little-spacer-left" />
        </a>
        {this.renderDropdown()}
        {this.renderMergeBranch()}
      </div>
    );
  }
}
