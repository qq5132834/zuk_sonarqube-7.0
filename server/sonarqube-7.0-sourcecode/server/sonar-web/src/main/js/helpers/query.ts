/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { isNil, omitBy } from 'lodash';
import { isValidDate, parseDate, toNotSoISOString } from './dates';

export interface RawQuery {
  [x: string]: any;
}

function arraysEqual(a: RawQuery, b: RawQuery): boolean {
  if (a.length !== b.length) {
    return false;
  }
  for (let i = 0; i < a.length; i++) {
    if (a[i] !== b[i]) {
      return false;
    }
  }
  return true;
}

export function queriesEqual(a: RawQuery, b: RawQuery): boolean {
  const keysA = Object.keys(a);
  const keysB = Object.keys(b);

  if (keysA.length !== keysB.length) {
    return false;
  }

  return keysA.every(
    key =>
      Array.isArray(a[key]) && Array.isArray(b[key])
        ? arraysEqual(a[key], b[key])
        : a[key] === b[key]
  );
}

export function cleanQuery(query: RawQuery): RawQuery {
  return omitBy(query, isNil);
}

export function parseAsBoolean(value: string | undefined, defaultValue: boolean = true): boolean {
  return value === 'false' ? false : value === 'true' ? true : defaultValue;
}

export function parseAsDate(value?: string): Date | undefined {
  if (value) {
    const date = parseDate(value);
    if (isValidDate(date)) {
      return date;
    }
  }
  return undefined;
}

export function parseAsFacetMode(facetMode: string): string {
  return facetMode === 'debt' || facetMode === 'effort' ? 'effort' : 'count';
}

export function parseAsString(value: string | undefined): string {
  return value || '';
}

export function parseAsArray(
  value: string | undefined,
  itemParser: (x: string) => string
): string[] {
  return value ? value.split(',').map(itemParser) : [];
}

export function serializeDate(value?: Date): string | undefined {
  if (value != null && value.toISOString) {
    return toNotSoISOString(value);
  }
  return undefined;
}

export function serializeString(value: string): string | undefined {
  return value || undefined;
}

export function serializeStringArray(value: string[] | undefined[]): string | undefined {
  return value && value.length ? value.join() : undefined;
}
