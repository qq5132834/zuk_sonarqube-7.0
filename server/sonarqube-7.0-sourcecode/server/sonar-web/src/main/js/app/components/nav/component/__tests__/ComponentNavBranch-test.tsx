/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import { shallow } from 'enzyme';
import ComponentNavBranch from '../ComponentNavBranch';
import {
  BranchType,
  ShortLivingBranch,
  MainBranch,
  Component,
  LongLivingBranch
} from '../../../../types';
import { click } from '../../../../../helpers/testUtils';

const fooBranch: LongLivingBranch = { isMain: false, name: 'foo', type: BranchType.LONG };

it('renders main branch', () => {
  const branch: MainBranch = { isMain: true, name: 'master' };
  const component = {} as Component;
  expect(
    shallow(
      <ComponentNavBranch
        branches={[branch, fooBranch]}
        component={component}
        currentBranch={branch}
      />,
      { context: { branchesEnabled: true } }
    )
  ).toMatchSnapshot();
});

it('renders short-living branch', () => {
  const branch: ShortLivingBranch = {
    isMain: false,
    mergeBranch: 'master',
    name: 'foo',
    status: { bugs: 0, codeSmells: 0, vulnerabilities: 0 },
    type: BranchType.SHORT
  };
  const component = {} as Component;
  expect(
    shallow(
      <ComponentNavBranch
        branches={[branch, fooBranch]}
        component={component}
        currentBranch={branch}
      />,
      { context: { branchesEnabled: true } }
    )
  ).toMatchSnapshot();
});

it('opens menu', () => {
  const branch: MainBranch = { isMain: true, name: 'master' };
  const component = {} as Component;
  const wrapper = shallow(
    <ComponentNavBranch
      branches={[branch, fooBranch]}
      component={component}
      currentBranch={branch}
    />,
    { context: { branchesEnabled: true } }
  );
  expect(wrapper.find('ComponentNavBranchesMenu')).toHaveLength(0);
  click(wrapper.find('a'));
  expect(wrapper.find('ComponentNavBranchesMenu')).toHaveLength(1);
});

it('renders single branch popup', () => {
  const branch: MainBranch = { isMain: true, name: 'master' };
  const component = {} as Component;
  const wrapper = shallow(
    <ComponentNavBranch branches={[branch]} component={component} currentBranch={branch} />,
    { context: { branchesEnabled: true } }
  );
  expect(wrapper).toMatchSnapshot();
  expect(wrapper.find('BubblePopupHelper').prop('isOpen')).toBe(false);
  click(wrapper.find('a'));
  expect(wrapper.find('BubblePopupHelper').prop('isOpen')).toBe(true);
});

it('renders no branch support popup', () => {
  const branch: MainBranch = { isMain: true, name: 'master' };
  const component = {} as Component;
  const wrapper = shallow(
    <ComponentNavBranch
      branches={[branch, fooBranch]}
      component={component}
      currentBranch={branch}
    />,
    { context: { branchesEnabled: false } }
  );
  expect(wrapper).toMatchSnapshot();
  expect(wrapper.find('BubblePopupHelper').prop('isOpen')).toBe(false);
  click(wrapper.find('a'));
  expect(wrapper.find('BubblePopupHelper').prop('isOpen')).toBe(true);
});

it('renders nothing on SonarCloud without branch support', () => {
  const branch: MainBranch = { isMain: true, name: 'master' };
  const component = {} as Component;
  const wrapper = shallow(
    <ComponentNavBranch branches={[branch]} component={component} currentBranch={branch} />,
    { context: { branchesEnabled: false, onSonarCloud: true } }
  );
  expect(wrapper.type()).toBeNull();
});
