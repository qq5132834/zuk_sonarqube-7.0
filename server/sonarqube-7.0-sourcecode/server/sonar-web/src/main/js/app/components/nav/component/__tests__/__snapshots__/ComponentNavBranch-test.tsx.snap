// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders main branch 1`] = `
<div
  className="navbar-context-branches dropdown"
>
  <a
    className="link-base-color link-no-underline"
    href="#"
    onClick={[Function]}
  >
    <BranchIcon
      branch={
        Object {
          "isMain": true,
          "name": "master",
        }
      }
      className="little-spacer-right"
    />
    <Tooltip
      mouseEnterDelay={1}
      overlay="master"
      placement="bottom"
    >
      <span
        className="text-limited text-top"
      >
        master
      </span>
    </Tooltip>
    <i
      className="icon-dropdown little-spacer-left"
    />
  </a>
</div>
`;

exports[`renders no branch support popup 1`] = `
<div
  className="navbar-context-branches"
>
  <BranchIcon
    branch={
      Object {
        "isMain": true,
        "name": "master",
      }
    }
    className="little-spacer-right"
    fill="#cdcdcd"
  />
  <span
    className="note"
  >
    master
  </span>
  <div
    className="display-inline-block spacer-left"
  >
    <a
      className="link-no-underline"
      href="#"
      onClick={[Function]}
    >
      <HelpIcon
        fill="#cdcdcd"
      />
    </a>
    <BubblePopupHelper
      isOpen={false}
      popup={<NoBranchSupportPopup />}
      position="bottomleft"
      togglePopup={[Function]}
    />
  </div>
</div>
`;

exports[`renders short-living branch 1`] = `
<div
  className="navbar-context-branches dropdown"
>
  <a
    className="link-base-color link-no-underline"
    href="#"
    onClick={[Function]}
  >
    <BranchIcon
      branch={
        Object {
          "isMain": false,
          "mergeBranch": "master",
          "name": "foo",
          "status": Object {
            "bugs": 0,
            "codeSmells": 0,
            "vulnerabilities": 0,
          },
          "type": "SHORT",
        }
      }
      className="little-spacer-right"
    />
    <Tooltip
      mouseEnterDelay={1}
      overlay="foo"
      placement="bottom"
    >
      <span
        className="text-limited text-top"
      >
        foo
      </span>
    </Tooltip>
    <i
      className="icon-dropdown little-spacer-left"
    />
  </a>
  <span
    className="note big-spacer-left text-lowercase"
  >
    from
     
    <strong>
      master
    </strong>
  </span>
</div>
`;

exports[`renders single branch popup 1`] = `
<div
  className="navbar-context-branches"
>
  <BranchIcon
    branch={
      Object {
        "isMain": true,
        "name": "master",
      }
    }
    className="little-spacer-right"
  />
  <span
    className="note"
  >
    master
  </span>
  <div
    className="display-inline-block spacer-left"
  >
    <a
      className="link-no-underline"
      href="#"
      onClick={[Function]}
    >
      <HelpIcon
        fill="#4b9fd5"
      />
    </a>
    <BubblePopupHelper
      isOpen={false}
      popup={<SingleBranchHelperPopup />}
      position="bottomleft"
      togglePopup={[Function]}
    />
  </div>
</div>
`;
